components/places/src/import/ios_bookmarks.rs

```diff
@@ -0,0 +1,566 @@
+            CASE b.type
+                WHEN {ios_bookmark_type} THEN {bookmark_kind}
+                WHEN {ios_folder_type} THEN {folder_kind}
+                WHEN {ios_separator_type} THEN {separator_kind}
+                -- We filter out anything else when inserting into the stage table
+            END,
+            b.date_added,
+            b.title,
+            -- placeId
+            CASE WHEN b.bmkUri IS NULL
+            THEN NULL
+            ELSE (SELECT id FROM main.moz_places p
+                  WHERE p.url_hash = hash(b.bmkUri) AND p.url = b.bmkUri)
+            END,
+            b.keyword
+        FROM iosBookmarksStaging b",
+        bookmark_kind = SyncedBookmarkKind::Bookmark as u8,
+        folder_kind = SyncedBookmarkKind::Folder as u8,
+        separator_kind = SyncedBookmarkKind::Separator as u8,
+
+        ios_bookmark_type = IosBookmarkType::Bookmark as u8,
+        ios_folder_type = IosBookmarkType::Folder as u8,
+        ios_separator_type = IosBookmarkType::Separator as u8,
+
+    );
+}
+
+// This could be a const &str, but it fits better here.
+const POPULATE_MIRROR_STRUCTURE: &str = "
+REPLACE INTO main.moz_bookmarks_synced_structure(guid, parentGuid, position)
+    SELECT structure.child, structure.parent, structure.idx FROM ios.bookmarksBufferStructure structure
+    WHERE EXISTS(
+        SELECT 1 FROM iosBookmarksStaging stage
+        WHERE stage.isLocal = 0
+            AND stage.guid = structure.child
+    );
+REPLACE INTO main.moz_bookmarks_synced_structure(guid, parentGuid, position)
+    SELECT structure.child, structure.parent, structure.idx FROM ios.bookmarksLocalStructure structure
+    WHERE EXISTS(
+        SELECT 1 FROM iosBookmarksStaging stage
+        WHERE stage.isLocal != 0
+            AND stage.guid = structure.child
+    );
+";
+
+lazy_static::lazy_static! {
+    static ref POPULATE_STAGING: String = format!(
+        "INSERT OR IGNORE INTO temp.iosBookmarksStaging(
+            guid,
+            type,
+            parentid,
+            pos,
+            title,
+            bmkUri,
+            keyword,
+            tags,
+            date_added,
+            modified,
+            isLocal
+        )
+        SELECT
+            b.guid,
+            b.type,
+            b.parentid,
+            b.pos,
+            b.title,
+            CASE
+                WHEN b.bmkUri IS NOT NULL
+                    THEN validate_url(b.bmkUri)
+                ELSE NULL
+            END,
+            b.keyword,
+            b.tags,
+            sanitize_timestamp(b.date_added),
+            sanitize_timestamp(b.server_modified),
+            0
+        FROM ios.bookmarksBuffer b
+        WHERE NOT b.is_deleted
+            -- Skip anything also in `local` (we can't use `replace`,
+            -- since we use `IGNORE` to avoid inserting bad records)
+            AND (
+                (b.guid IN {roots})
+                OR
+                (b.guid NOT IN (SELECT l.guid FROM ios.bookmarksLocal l))
```

Comment:
I just tried this and it seems to slightly slow down things, actually (315ms vs 340ms for importing my tree of 1000 bookmarks). Possibly worth noting that local will typically be much smaller than remote when they both exist.

Also I think it reads a bit more clearly this way, if only because you don't see LEFT JOIN every day.

Reply:
Good to know, thanks for double-checking!